File: agent/checkpointer.js
// Phase 3b — checkpointer selection.
//
// Returns the LangGraph checkpointer that gives the graph its conversation
// memory (keyed by thread_id):
//   - Postgres (durable, survives restarts) when DATABASE_URL is set, or
//   - in-memory MemorySaver (lost on restart) as a fallback.
//
// The Postgres saver needs a one-time .setup() to create its checkpoint tables;
// we run it here so callers just get a ready-to-use checkpointer.

import { MemorySaver } from "@langchain/langgraph";

/**
 * @returns {Promise<{ checkpointer: import("@langchain/langgraph").BaseCheckpointSaver, kind: "postgres"|"memory", detail: string }>}
 */
export async function getCheckpointer() {
  const url = process.env.DATABASE_URL;
  if (!url) {
    return { checkpointer: new MemorySaver(), kind: "memory", detail: "in-memory (no DATABASE_URL)" };
  }

  const { PostgresSaver } = await import("@langchain/langgraph-checkpoint-postgres");
  const saver = PostgresSaver.fromConnString(url);
  await saver.setup(); // create checkpoint tables if they don't exist yet

  // Show host/db without leaking credentials.
  let detail = "postgres";
  try {
    const u = new URL(url);
    detail = `postgres ${u.host}${u.pathname}`;
  } catch {
    /* ignore */
  }
  return { checkpointer: saver, kind: "postgres", detail };
}

File: server/app/state/checkpointer.py
"""Checkpointer selection (reference: agent/checkpointer.js).

Postgres (durable, survives restarts) when DATABASE_URL is set, else an in-memory
saver (lost on restart). The Postgres saver owns a connection, so it is opened as
an async context and kept open for the app lifetime (FastAPI lifespan).
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from dataclasses import dataclass
from urllib.parse import urlsplit

from langgraph.checkpoint.base import BaseCheckpointSaver
from langgraph.checkpoint.memory import InMemorySaver


@dataclass(frozen=True)
class Checkpointer:
    saver: BaseCheckpointSaver
    kind: str  # "postgres" | "memory"
    detail: str


def _describe(url: str) -> str:
    """host/db without credentials."""
    try:
        u = urlsplit(url)
        return f"postgres {u.hostname}:{u.port or 5432}{u.path}"
    except ValueError:
        return "postgres"


@asynccontextmanager
async def open_checkpointer(database_url: str | None) -> AsyncIterator[Checkpointer]:
    if not database_url:
        yield Checkpointer(InMemorySaver(), "memory", "in-memory (no DATABASE_URL)")
        return

    from langgraph.checkpoint.postgres.aio import AsyncPostgresSaver

    async with AsyncPostgresSaver.from_conn_string(database_url) as saver:
        await saver.setup()  # create checkpoint tables if they don't exist yet
        yield Checkpointer(saver, "postgres", _describe(database_url))
